### ISSUE TITLE
Question about loss calculation in `examples/table2text/trainer.py `

### ISSUE THREAD

**Issue by jerryyiransun** [awslabs/fast-differential-privacy#48]
I have tried to turn on logging in `examples/table2text/run.sh b/examples/table2text/run.sh` by changing `--logging_steps` to 1. But the loss I am getting from the training looks like:

```
{'loss': 1257.6123046875, 'learning_rate': 0.002, 'epoch': 0.023784606602606794, 'total_flos': 52080157753344, 'step': 1, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 2634.69140625, 'learning_rate': 0.002, 'epoch': 0.04756921320521359, 'total_flos': 104279023484928, 'step': 2, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1867.05615234375, 'learning_rate': 0.002, 'epoch': 0.07135381980782038, 'total_flos': 156819174653952, 'step': 3, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1691.029296875, 'learning_rate': 0.002, 'epoch': 0.09513842641042718, 'total_flos': 208943847899136, 'step': 4, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1430.47607421875, 'learning_rate': 0.002, 'epoch': 0.11892303301303396, 'total_flos': 261608642445312, 'step': 5, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 2163.625, 'learning_rate': 0.002, 'epoch': 0.14270763961564076, 'total_flos': 313546350624768, 'step': 6, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1554.810546875, 'learning_rate': 0.002, 'epoch': 0.16649224621824754, 'total_flos': 365478123405312, 'step': 7, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1390.923828125, 'learning_rate': 0.002, 'epoch': 0.19027685282085435, 'total_flos': 417641376743424, 'step': 8, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1284.138671875, 'learning_rate': 0.002, 'epoch': 0.21406145942346114, 'total_flos': 470398169972736, 'step': 9, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1199.9453125, 'learning_rate': 0.002, 'epoch': 0.23784606602606792, 'total_flos': 523374572961792, 'step': 10, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1111.04296875, 'learning_rate': 0.002, 'epoch': 0.26163067262867473, 'total_flos': 575564535595008, 'step': 11, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1098.80078125, 'learning_rate': 0.002, 'epoch': 0.2854152792312815, 'total_flos': 627347923402752, 'step': 12, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1060.6796875, 'learning_rate': 0.002, 'epoch': 0.3091998858338883, 'total_flos': 679609110822912, 'step': 13, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 1033.302734375, 'learning_rate': 0.002, 'epoch': 0.3329844924364951, 'total_flos': 732155197390848, 'step': 14, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 981.1796875, 'learning_rate': 0.002, 'epoch': 0.35676909903910187, 'total_flos': 784434191007744, 'step': 15, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 950.671875, 'learning_rate': 0.002, 'epoch': 0.3805537056417087, 'total_flos': 837345304608768, 'step': 16, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 909.5625, 'learning_rate': 0.002, 'epoch': 0.4043383122443155, 'total_flos': 889541202640896, 'step': 17, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 877.14453125, 'learning_rate': 0.002, 'epoch': 0.4281229188469223, 'total_flos': 941274139557888, 'step': 18, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
{'loss': 813.21484375, 'learning_rate': 0.002, 'epoch': 0.45190752544952906, 'total_flos': 993274169425920, 'step': 19, 'num_params': 163038720, 'num_non_embedding_params': 85056000}
```

I was looking at `examples/table2text/trainer.py`, specifically:
```
loss = F.cross_entropy(shift_logits.permute(0, 2, 1), shift_labels)
```
in `compute_loss`, and
```
loss = self.compute_loss(model, inputs)
if self.args.deepspeed_config:
    model.backward(loss)
else:
    loss.backward()
```
in `training_step`.

From my understanding:

- F.cross_entropy by default returns the mean over all non-ignored tokens in the local batch. In DDP/multi-GPU, that means each rank computes and logs a per-token mean for its shard. So the printed losses differ between GPUs and aren’t aggregated.

- In the non-DeepSpeed branch, the loss isn’t divided by gradient_accumulation_steps before backward(), which could scale gradients differently depending on the accumulation setting.

My questions are:
1. Is the current intention to log/optimize based on a local per-token mean loss per GPU, without cross-GPU averaging?
2. Should the non-DeepSpeed branch divide the loss by gradient_accumulation_steps before backward() to match standard gradient accumulation behavior?

**Comment by jerryyiransun** [comment]
@woodyx218 Hello Woody, I've been trying to get the examples running with proper losses but it seems like I am still getting something unreasonable. Do you have any recommendation on what I could do? Any help will be greatly appreciated! Thank you in advance.

**Comment by woodyx218** [comment]
The code is indeed doing the cross-GPU grad averaging correctly. You don't need to divide by grad accumulations if you use loss_reduction==sum or DeepSpeed version. It may be a package dependence issue or logging issue.

P.S. I'm no longer with Amazon and won't be able to oversee this repo.

**Comment by jerryyiransun** [comment]
Got it, I will look into my dependencies, thank you!
